### PR TITLE
fix: [CDS-75591]: fixed special chars secret masking for winrm

### PR DESCRIPTION
**Pull request by vitsafronovici**
## Describe your changes
Currently only for WinRm deployments special characters (<, >, &, |) are replaced with (^<, ^>, ^&, ^|) this is a powershell specific escaping in order to avoid reserved characters.
The problem comes if for example a secret has value `<ab>` and during escaping the output log line is `^<ab^>`. Sanitizing does not mask the escaped secret. In order to achieve that and to keep performance in mind, we need an extra operation that would take into account the escaped secrets also and only for WinRm deployments

Before:
image
After:
image

## Checklist
- [ ] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/51454)

### CODE REVIEW

**Review by ivan-83 — CHANGES_REQUESTED** [review]
LGTM with comments, the idea is good just a couple of comments. We need to add and delegate the team to review code changes.

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51454#-:-NbKBjhXEeg14eEBX8dI:b-j9opi5)* status: 7 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java` line 77 at r1](https://reviewable.io/reviews/harness/harness-core/51454#-NbK8wzX7vaOOXnAWms4:-NbK8wzX7vaOOXnAWms5:bavppm2) ([raw file](https://github.com/harness/harness-core/blob/0eb564954a274d84a6a2e121758ecbab7d19a432/260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java#L77)):*
> ```Java
>   @Default private final Map<String, List<LogLine>> logCache = new HashMap<>();
> 
>   private Map<String, Object> flags;
> ```

This is the concept of Marker. What I suggested is having HashSet<> markers; instead of Map<String, Object> flags
___
*[`260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java` line 140 at r1](https://reviewable.io/reviews/harness/harness-core/51454#-NbK9IBm0xmlBci7muid:-NbK9IBnEETXEQZxP00z:b-va4h1) ([raw file](https://github.com/harness/harness-core/blob/0eb564954a274d84a6a2e121758ecbab7d19a432/260-delegate/src/main/java/io/harness/logstreaming/LogStreamingTaskClient.java#L140)):*
> ```Java
> 
>     boolean isWinRm = Boolean.TRUE.equals(getFlags().get(WIN_RM_FLAG));
>     logStreamingSanitizer.sanitizeLogMessage(logLine, isWinRm);
> ```

Make sanitizeLogMessage method reusable and pass markers instead of isWinRm. We can use HashSet.contains(WIN_RM_MARKER) in sanitizeLogMessage method. It's O(1) and totally fine for logger. 
___
*[`930-delegate-tasks/src/main/java/io/harness/delegate/task/shell/CommandTaskNG.java` line 91 at r1](https://reviewable.io/reviews/harness/harness-core/51454#-NbK9a7D8gFi_TKSGnkl:-NbK9a7D8gFi_TKSGnkm:b-wp2auk) ([raw file](https://github.com/harness/harness-core/blob/0eb564954a274d84a6a2e121758ecbab7d19a432/930-delegate-tasks/src/main/java/io/harness/delegate/task/shell/CommandTaskNG.java#L91)):*
> ```Java
> 
>   private DelegateResponseData runWinRm(WinrmTaskParameters parameters) {
>     getLogStreamingTaskClient().getFlags().put(WIN_RM_FLAG, true);
> ```

Here add to HashSet WIN_RM_MARKER
___
*[`950-log-client/src/main/java/io/harness/logstreaming/LogStreamingSanitizer.java` line 37 at r1](https://reviewable.io/reviews/harness/harness-core/51454#-NbKB38540cobfsd23Bj:-NbKB38540cobfsd23Bk:bzdxz6m) ([raw file](https://github.com/harness/harness-core/blob/0eb564954a274d84a6a2e121758ecbab7d19a432/950-log-client/src/main/java/io/harness/logstreaming/LogStreamingSanitizer.java#L37)):*
> ```Java
>   }
> 
>   public void sanitizeLogMessage(LogLine logLine, boolean isWinRm) {
> ```

The second approach could be instead of boolean isWinRM just pass Markers without modifying LogStreamingTaskClient class.

**Comment by vitsafronovici** [comment]
trigger uts

**Comment by vitsafronovici** [comment]
trigger uts

**Review by vitsafronovici — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51454)* status: 8 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. Waiting on team(s): del-team. 

___
*[`950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/logstreaming/ILogStreamingTaskClient.java` line 26 at r3](https://reviewable.io/reviews/harness/harness-core/51454#gh-1287296710:-NbO17tE5TMnTeJvwKhE:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/f9623eadb1b1f8f12da3b9605d67c1cfb32c3afc/950-delegate-tasks-beans/src/main/java/io/harness/delegate/beans/logstreaming/ILogStreamingTaskClient.java#L26)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

Move WIN_RM_MARKER to the common class. 
</blockquote></details>

done
___
*[`950-log-client/BUILD.bazel` line 25 at r3](https://reviewable.io/reviews/harness/harness-core/51454#gh-1287297576:-NbKrnNh2-E5cqt5dTm-:bdowcq1) ([raw file](https://github.com/harness/harness-core/blob/f9623eadb1b1f8f12da3b9605d67c1cfb32c3afc/950-log-client/BUILD.bazel#L25)):*
<details><summary><i>Previously, acasian-harness (Alexandru Casian) wrote…</i></summary><blockquote>

what is this module for?
</blockquote></details>

import io.harness.windows.CmdUtils;
___
*[`950-log-client/src/main/java/io/harness/logstreaming/LogStreamingSanitizer.java` line 33 at r3](https://reviewable.io/reviews/harness/harness-core/51454#gh-1287299302:-NbO1A54-iYGFisQW-tD:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/f9623eadb1b1f8f12da3b9605d67c1cfb32c3afc/950-log-client/src/main/java/io/harness/logstreaming/LogStreamingSanitizer.java#L33)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

Agree. We can move it to another utility class.
</blockquote></details>

Done.

**Comment by vitsafronovici** [comment]
trigger ut0, ut8

**Comment by vitsafronovici** [comment]
trigger ut6

**Review by ivan-83 — DISMISSED** [review]
> Reviewed 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51454)* status: 8 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained. Waiting on team(s): del-team.

**Review by acasian-harness — APPROVED** [review]

File: 930-delegate-tasks/BUILD.bazel [hunk 1]
```diff
@@ -260,6 +260,7 @@ java_library(
         "//933-ci-commons:module",
         "//940-secret-manager-client:module",
         "//950-delegate-tasks-beans:module",
+        "//950-log-client:module",
```
Comment (jennyjames21) [inline 930-delegate-tasks/BUILD.bazel:263]: Do we need a new module for 2 files?

File: 950-log-client/src/main/java/io/harness/logstreaming/LogStreamingSanitizer.java [hunk 2]
```diff
@@ -22,23 +26,38 @@
...
+  public void sanitizeLogMessage(LogLine logLine, Set<String> markers) {
     String sanitizedLogMessage = logLine.getMessage();
     if (isEmpty(sanitizedLogMessage)) {
       logLine.setMessage(sanitizedLogMessage);
       return;
     }
+
     if (!isEmpty(secrets)) {
+      boolean isWinRm = markers != null && markers.contains(WIN_RM_MARKER);
+      Set<String> allSecrets = isWinRm ? new HashSet<>(secrets) : secrets;
```
Comment (jennyjames21) [inline 950-log-client/src/main/java/io/harness/logstreaming/LogStreamingSanitizer.java:53]: why new HashSet here?

Comment (vitsafronovici) [reply 950-log-client/src/main/java/io/harness/logstreaming/LogStreamingSanitizer.java:53]: In order to add items
`Collectors.toCollection(() -> allSecrets`

**Review by vitsafronovici — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51454)* status: 8 file(s) to review. 4. unresolved discussions. 

___
*[`930-delegate-tasks/BUILD.bazel` line 263 at r4](https://reviewable.io/reviews/harness/harness-core/51454#gh-1288854963:-NbTbXGe1UR_wcx09SSn:b-rt0f3k) ([raw file](https://github.com/harness/harness-core/blob/655e4df6b7759101663305dd610de643622a2ce4/930-delegate-tasks/BUILD.bazel#L263)):*
<details><summary><i>Previously, jennyjames21 (Jenny James) wrote…</i></summary><blockquote>

Do we need a new module for 2 files? 
</blockquote></details>

this module has only one package so there is no need to add bazel build file for that package separately

**Comment by vitsafronovici** [comment]
trigger uts

**Review by vitsafronovici — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51454)* status: 9 file(s) to review. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-delegate-tasks/BUILD.bazel` line 263 at r4](https://reviewable.io/reviews/harness/harness-core/51454#gh-1288854963:-NbU-VGnBpnKnXsrU_8T:b8bf0ld) ([raw file](https://github.com/harness/harness-core/blob/655e4df6b7759101663305dd610de643622a2ce4/930-delegate-tasks/BUILD.bazel#L263)):*
<details><summary><i>Previously, vitsafronovici (Vitalie Safronovici) wrote…</i></summary><blockquote>

this module has only one package so there is no need to add bazel build file for that package separately
</blockquote></details>

ok, moved constant to 980-commons

**Review by ivan-83 — DISMISSED** [review]
:lgtm:

> Reviewed 6 of 6 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51454#-:-NbU1eZx8Pz4Xa_uak-N:bnfp4nl)* status: 9 file(s) to review. 4. unresolved discussions. 1 of 2 approvals obtained.

**Comment by vitsafronovici** [comment]
trigger uts

**Review by acasian-harness — APPROVED** [review]

**Review by jennyjames21 — APPROVED** [review]

**Comment by vitsafronovici** [comment]
trigger review

**Pull request merged**